Mathlib/Algebra/Module/Injective.lean

```diff
@@ -39,7 +39,7 @@ import Mathlib.Algebra.Module.ULift
 
 noncomputable section
 
-universe u v
+universe u v v'
 
 variable (R : Type u) [Ring R] (Q : TypeMax.{v,u}) [AddCommGroup Q] [Module R Q]
```

Comment:
I think there's no reason `Q` can't be in an arbitrary universe; the test objects `X` and `Y` need to be in the universe of `R` (or higher, but Module.Baer is purely in the universe of `R`), but `Q` need not be. This seems to have [made it hard to use](https://leanprover.zulipchat.com/#narrow/stream/116395-maths/topic/Universe.20trouble.20when.20using.20.60.60.60Module.2EInjective.60.60.60/near/403006687). Can you address this issue as well?

It's an interesting question to ponder whether it suffices to use test objects in the same universe as `Q`; I haven't been able to come up with counterexamples or proof.

Reply:
For the universe problem, I tried 3 things

1. Test object in higher universe:
```lean
class Module.Injective : Prop where
  out : ∀ ⦃X Y : Type max u v⦄ [AddCommGroup X] [AddCommGroup Y] [Module R X] [Module R Y]
    (f : X →ₗ[R] Y) (_ : Function.Injective f) (g : X →ₗ[R] Q),
    ∃ h : Y →ₗ[R] Q, ∀ x, h (f x) = g x
```

```lean
theorem Module.injective_object_of_injective_module [inj : Module.Injective R Q] :
    CategoryTheory.Injective (ModuleCat.of R Q) where
  factors {X Y} g f m := by
    let f' : ULift.{max u v} X →ₗ[R] ULift.{max u v} Y :=
      ULift.moduleEquiv.symm.toLinearMap ∘ₗ f ∘ₗ ULift.moduleEquiv.toLinearMap
    have hf' : Function.Injective f' := ULift.up_injective.comp <|
      (ModuleCat.mono_iff_injective _).mp m |>.comp ULift.down_injective
    let g' : ULift.{max u v} X →ₗ[R] Q := g ∘ₗ ULift.moduleEquiv.toLinearMap
    obtain ⟨l, h⟩ := inj.out f' hf' g'
    exact ⟨l ∘ₗ ULift.moduleEquiv.symm.toLinearMap, LinearMap.ext fun x ↦ by aesop⟩
```
is doable by playing with `ULift.moduleEquiv`, but
```lean
theorem Module.injective_module_of_injective_object
    [inj : CategoryTheory.Injective <| ModuleCat.of R Q] :
    Module.Injective R Q where
  out X Y _ _ _ _ f hf g := by
    have H := @CategoryTheory.Injective.factors (self := inj)
    sorry
```
seems not to be possible because `X, Y` are in universe `max u v` but `H` want universe modules to be in universe `v`


2. `Q, X, Y` all in the same universe

```lean
class Module.Injective : Prop where
  out : ∀ ⦃X Y : Type v⦄ [AddCommGroup X] [AddCommGroup Y] [Module R X] [Module R Y]
    (f : X →ₗ[R] Y) (_ : Function.Injective f) (g : X →ₗ[R] Q),
    ∃ h : Y →ₗ[R] Q, ∀ x, h (f x) = g x
```
is fine up until `Module.Baer.of_injective` because ideals are in universe `u` and can't be easily transferred to `v`.

3. use `[UnivLE.{u, v}]` to solve the problem in the second approach, now we can `Shrink I` to make an object in universe `v`, but I don't think it can be shown easily, if at all, that `Shrink I ≃ₗ[R] I`

Reply:
I think your solution 1 is probably the best for now; you can have a version of `Module.injective_module_of_injective_object` that uses `ULift Q` and another version that requires `Q : TypeMax`.

Reply:
In [9e067bb](https://github.com/leanprover-community/mathlib4/pull/8905/commits/9e067bb375f91e015774651c4f725f89b5956be5), I made a version of solution 3, it is actually quite smooth. Does this change your preference of solution 1?

Reply:
It doesn't build yet? Is it really solution 3, I don't see any UnivLE in it?

Reply:
I forgot to push and quoted the wrong commit. The correct one is [10a89bc](https://github.com/leanprover-community/mathlib4/pull/8905/commits/10a89bc7b4794ae0a4ccb210f31c1f3ea7bbcb82) :laughing:

Reply:
The full version using `UnivLE` is in [0ce04e6](https://github.com/leanprover-community/mathlib4/pull/8905/commits/0ce04e69b49037aea05fa7b13dff8d6d70237deb)

Reply:
I really think `Module.injective_module_iff_injective_object` being a provable statement without lifting `Q` should be a consideration when assessing what the correct `Module.Injective` version should be.

Reply:
I agree! See https://github.com/leanprover-community/mathlib4/pull/8905#discussion_r1428509361 for some ideas.